lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -209,26 +300,267 @@ classifyCodingKeys(const DerivedConformance &derived) {
+
+  // Only ill-formed code would produce multiple results for this lookup.
+  // This would get diagnosed later anyway, so we're free to only look at the
+  // first result here.
+  EnumDecl *codingKeysEnum = nullptr;
+  if (!needsSynthesis) {
+    codingKeysEnum =
+        validateCodingKeysProtocolConformance(derived, codingKeysDecls.front());
+    if (!codingKeysEnum) {
+      return CodingKeysClassification::Invalid;
+    }
+  }
+
+  if (auto *enumDecl = dyn_cast<EnumDecl>(derived.Nominal)) {
+    for (auto *elt : enumDecl->getAllElements()) {
+      Identifier caseCodingKeyId = caseCodingKeyIdentifier(C, elt);
+      auto caseCodingKeysDecls =
+          derived.Nominal->lookupDirect(DeclName(caseCodingKeyId));
+      if (caseCodingKeysDecls.empty()) {
+        needsSynthesis = true;
+      } else {
+        auto *caseCodingKeysEnum = validateCodingKeysProtocolConformance(
+            derived, caseCodingKeysDecls.front());
+        if (!caseCodingKeysEnum) {
+          return CodingKeysClassification::Invalid;
+        }
+
+        if (!validateCaseCodingKeysEnum(derived, elt, caseCodingKeysEnum)) {
+          return CodingKeysClassification::Invalid;
+        }
+      }
+    }
+    return needsSynthesis ? CodingKeysClassification::NeedsSynthesizedCodingKeys
+                          : CodingKeysClassification::Valid;
+  } else if (needsSynthesis) {
+    return CodingKeysClassification::NeedsSynthesizedCodingKeys;
+  } else {
+    return validateCodingKeysEnum(derived, codingKeysEnum)
+               ? CodingKeysClassification::Valid
+               : CodingKeysClassification::Invalid;
+  }
+}
+
+/// Fetches the \c CodingKeys enum nested in \c target, potentially reaching
+/// through a typealias if the "CodingKeys" entity is a typealias.
+///
+/// This is only useful once a \c CodingKeys enum has been validated (via \c
+/// hasValidCodingKeysEnum) or synthesized (via \c synthesizeCodingKeysEnum).
+///
+/// \param C The \c ASTContext to perform the lookup in.
+///
+/// \param target The target type to look in.
+///
+/// \return A retrieved canonical \c CodingKeys enum if \c target has a valid
+/// one; \c nullptr otherwise.
+static EnumDecl *lookupEvaluatedCodingKeysEnum(ASTContext &C,
+                                               NominalTypeDecl *target) {
+  auto codingKeyDecls = target->lookupDirect(DeclName(C.Id_CodingKeys));
+  if (codingKeyDecls.empty())
+    return nullptr;
+
+  auto *codingKeysDecl = codingKeyDecls.front();
+  if (auto *typealiasDecl = dyn_cast<TypeAliasDecl>(codingKeysDecl))
+    codingKeysDecl = typealiasDecl->getDeclaredInterfaceType()->getAnyNominal();
+
+  return dyn_cast<EnumDecl>(codingKeysDecl);
+}
+
+static EnumDecl *lookupEvaluatedCodingKeysEnum(ASTContext &C,
+                                               NominalTypeDecl *target,
+                                               Identifier identifier) {
+  auto codingKeyDecls = target->lookupDirect(DeclName(identifier));
```

Comment:
Interesting quirk: Because you're using `lookupDirect()` instead of a qualified lookup, IIRC you won't be looking in protocol extensions you conform to. I'm not sure this is *wrong*, mind you, just something to be aware of.

Reply:
Interesting. The version that does not take an `Identifier` is actually not new, it just moved around. I don't know why it is this way.